File: BHAVRP_Java/BHAVRPv2.0/BHAVRP2.0/src/cujae/inf/ic/om/assignment/clustering/partitional/ByMedoids.java
package cujae.inf.ic.om.assignment.clustering.partitional;

import java.util.ArrayList;
import java.util.Random;

import cujae.inf.ic.om.factory.DistanceType;
import cujae.inf.ic.om.assignment.clustering.SamplingType;
import cujae.inf.ic.om.matrix.NumericMatrix;
import cujae.inf.ic.om.matrix.RowCol;
import cujae.inf.ic.om.problem.input.Customer;
import cujae.inf.ic.om.problem.input.Depot;
import cujae.inf.ic.om.problem.input.Fleet;
import cujae.inf.ic.om.problem.input.Location;
import cujae.inf.ic.om.problem.input.Problem;
import cujae.inf.ic.om.problem.solution.Cluster;

public abstract class ByMedoids extends Partitional {
	
	// generar particion secuencial y aleatoria
	protected ArrayList<ArrayList<Customer>> generatePartitions(int sampsize, SamplingType samplingType){
		ArrayList<ArrayList<Customer>> partitions = new ArrayList<ArrayList<Customer>>();
		ArrayList<Customer> partition = new ArrayList<Customer>();
		
		int totalCustomers = Problem.getProblem().getTotalCustomers();
		int totalPartitions = totalCustomers/sampsize;
		
		if(totalCustomers%sampsize != 0)
			totalPartitions += 1;
			
		System.out.println("TOTAL DE PARTICIONES: " + totalPartitions);
		System.out.println("---------------------------------------------------------------");
		
	
		switch(samplingType.ordinal())
		{
			case 0:
			{
				int j = 0;
				int posElement = -1;
				Random rdm = new Random();
				
				ArrayList<Customer> customers = new ArrayList<Customer>(Problem.getProblem().getCustomers());
				
				for(int i = 0; i < totalPartitions; i++)
				{
					while((j < (sampsize * (i + 1))) && (j < totalCustomers))
					{
						posElement = rdm.nextInt(customers.size());
						partition.add(customers.remove(posElement));
						
						j++;
					}
					
					System.out.println("PARTICIÓN " + (i + 1) + ": ");
					System.out.println("TOTAL DE ELEMENTOS DE LA PARTICIÓN " + (i + 1) + ": " + partition.size());
					for(int k = 0; k < partition.size(); k++)
						System.out.println("ELEMENTOS DE LA PARTICIÓN " + (i + 1) + ": " +  partition.get(k).getIDCustomer());
					System.out.println("---------------------------------------------------------------");
					
					partitions.add(partition);
					partition = new ArrayList<Customer>();
				}
				
				break;
			}
			case 1:
			{
				int j = 0;
				
				for(int i = 0; i < totalPartitions; i++)
				{
					while((j < (sampsize * (i + 1))) && (j < totalCustomers))
					{
						partition.add(Problem.getProblem().getCustomers().get(j));
						j++;
					}
					
					System.out.println("PARTICIÓN " + (i + 1) + ": ");
					System.out.println("TOTAL DE ELEMENTOS DE LA PARTICIÓN " + (i + 1) + ": " + partition.size());
					for(int k = 0; k < partition.size(); k++)
						System.out.println("ELEMENTOS DE LA PARTICIÓN " + (i + 1) + ": " +  partition.get(k).getIDCustomer());
					System.out.println("---------------------------------------------------------------");
					
					partitions.add(partition);
					partition = new ArrayList<Customer>();
				}
				
				break;
			}
		}
		
		return partitions;
	}
	
	protected ArrayList<Integer> generateElements(ArrayList<Customer> customers, DistanceType distanceType){
		ArrayList<Integer> idElements = new ArrayList<Integer>();
		
		int totalCustomers = customers.size();
		int totalDepots = Problem.getProblem().getTotalDepots();
		int counter = totalDepots;
		
		NumericMatrix costMatrix = initializeCostMatrix(customers, Problem.getProblem().getDepots(), distanceType);
		RowCol rcBestAll = new RowCol();
		int idElement = -1 ;
		
		for(int i = 0; i < counter; i++)
			idElements.add(-1);

		System.out.println("LISTADO DE ELEMENTOS SELECCIONADOS:" + idElements);

		while(counter > 0)
		{
			rcBestAll = costMatrix.indexLowerValue(totalCustomers, 0, (totalCustomers + totalDepots - 1), (totalCustomers - 1));

			System.out.println("ROW SELECCIONADA: " + rcBestAll.getRow());
			System.out.println("COL SELECCIONADA: " + rcBestAll.getCol());
			System.out.println("VALOR SELECCIONADO: " + costMatrix.getItem(rcBestAll.getRow(), rcBestAll.getCol()));

			idElement = customers.get(rcBestAll.getCol()).getIDCustomer();	
			idElements.set((rcBestAll.getRow() - totalCustomers), idElement);	

			System.out.println("ELEMENTO: " + idElement); 
			System.out.println("LISTADO DE ELEMENTOS ACTUALIZADOS:" + idElements);

			costMatrix.fillValue(totalCustomers, rcBestAll.getCol(), (totalCustomers + totalDepots - 1), rcBestAll.getCol(), Double.POSITIVE_INFINITY);
			costMatrix.fillValue(rcBestAll.getRow(), 0, rcBestAll.getRow(), (totalCustomers + totalDepots - 1), Double.POSITIVE_INFINITY);
			counter--;
		}

	System.out.println("--------------------------------------------------");
	System.out.println("CENTROIDES/MEDOIDES INICIALES");
	System.out.println(idElements);
	System.out.println("--------------------------------------------------");

	return idElements;
}
	
	protected ArrayList<Integer> getIDMedoids(ArrayList<Depot> medoids){
		ArrayList<Integer> idMedoids = new ArrayList<Integer>();
		
		for(int i = 0; i < medoids.size(); i++)
			idMedoids.add(medoids.get(i).getIDDepot());
		
		System.out.println("--------------------------------------------------");
		System.out.println("ID MEDOIDES ACTUALES");
		System.out.println("--------------------------------------------------");
		System.out.println(idMedoids);
		
		return idMedoids;			
	}
	
	protected void updateClusters(ArrayList<Cluster> clusters, ArrayList<Integer> idElements) {
		
		for(int i = 0; i < clusters.size(); i++)
		{
			clusters.get(i).getItemsOfCluster().add(idElements.get(i));
			clusters.get(i).setRequestCluster(Problem.getProblem().getRequestByIDCustomer(idElements.get(i)));
		}
		
		System.out.println("--------------------------------------------------");
		System.out.println("LISTA DE CLUSTERS");
		for(int i = 0; i < clusters.size(); i++)
		{
			System.out.println("ID CLUSTER: " + clusters.get(i).getIDCluster());
			System.out.println("DEMANDA DEL CLUSTER: " + clusters.get(i).getRequestCluster());
			System.out.println("ELEMENTOS DEL CLUSTER: " + clusters.get(i).getItemsOfCluster());
		}
		System.out.println("--------------------------------------------------");
	}
	
	protected boolean verifyMedoids(ArrayList<Depot> oldMedoids, ArrayList<Depot> currentMedoids) {
		boolean change = false;
		int i = 0;
		
		while((!change) && (i < currentMedoids.size()))
		{
			if((oldMedoids.get(i).getLocationDepot().getAxisX() != currentMedoids.get(i).getLocationDepot().getAxisX()) || (oldMedoids.get(i).getLocationDepot().getAxisY() != currentMedoids.get(i).getLocationDepot().getAxisY()))
				change = true;
			else
				i++;
		}
		
		System.out.println("change:  " + change);
		
		return change;
	}
	
	protected ArrayList<Depot> replicateDepots(ArrayList<Depot> depots){
		ArrayList<Depot> newDepots = new ArrayList<Depot>();
		
		System.out.println("--------------------------------------------------");
		System.out.println("MEDOIDES/CENTROIDES ACTUALES");
		
		for(int i = 0; i < depots.size(); i++)
		{
			Depot depot = new Depot();
			
			depot.setIDDepot(depots.get(i).getIDDepot());
			
			double axisX = 0.0; 
			double axisY = 0.0; 
			axisX = depots.get(i).getLocationDepot().getAxisX();
			axisY = depots.get(i).getLocationDepot().getAxisY();
			
			Location location = new Location();
			location.setAxisX(axisX);
			location.setAxisY(axisY);
			depot.setLocationDepot(location);
			
			ArrayList<Fleet> fleet = new ArrayList<Fleet>();
			fleet.addAll(Problem.getProblem().getDepots().get(i).getFleetDepot());
			depot.setFleetDepot(fleet);
			
			newDepots.add(depot);
			
			System.out.println("--------------------------------------------------");
			System.out.println("ID MEDOIDE/CENTROIDE: " + newDepots.get(i).getIDDepot());
			System.out.println("X: " + newDepots.get(i).getLocationDepot().getAxisX());
			System.out.println("Y: " + newDepots.get(i).getLocationDepot().getAxisY());
			System.out.println("CAPACIDAD DE VEHICULO: " + newDepots.get(i).getFleetDepot().get(0).getCapacityVehicle());
			System.out.println("CANTIDAD DE VEHICULOS: " + newDepots.get(i).getFleetDepot().get(0).getCountVehicles());
		}
		
		return newDepots;
	}

	protected double calculateDissimilarity(DistanceType distanceType, ArrayList<Cluster> clusters) {
		double currentDissimilarity = 0.0;
		NumericMatrix dissimilarityMatrix = initializeCostMatrix(Problem.getProblem().getCustomers(), Problem.getProblem().getDepots(), distanceType);
		
		int posFirstItem = -1;
		int posSecondItem = -1;
		int totalClusters = clusters.size(); 
		int totalItems = 0; 
		
		for(int i = 0; i < totalClusters; i++)
		{
			totalItems = clusters.get(i).getItemsOfCluster().size(); 
			
			for(int j = 0; j < totalItems; j++)
			{
				posFirstItem = Problem.getProblem().getPosElement(clusters.get(i).getItemsOfCluster().get(j));
				
				for(int k = (j + 1); k < totalItems; k++)
				{
					posSecondItem = Problem.getProblem().getPosElement(clusters.get(i).getItemsOfCluster().get(k));
					currentDissimilarity += dissimilarityMatrix.getItem(posFirstItem, posSecondItem);
				}
			}
		}
		
		currentDissimilarity /= totalClusters;
		
		System.out.println("COEFICIENTE DE DISIMILITUD ACTUAL: " + currentDissimilarity);	
		
		return currentDissimilarity;
	}
}

File: BHAVRP_Python/cujae/inf/citi/om/assignment/clustering/partitional/by_medoids.py
import random
import numpy as np
from typing import List
from .partitional import Partitional
from ..sampling_type import SamplingType
from ....service.distance_type import DistanceType
from ....problem.input.problem import Problem
from ....problem.input.customer import Customer
from ....problem.input.depot import Depot
from ....problem.input.location import Location
from ....problem.solution.cluster import Cluster

class ByMedoids(Partitional):
    
    def __init__(self):
        super().__init__()
        
    # Método que genera particiones de clientes con tamaños especificados, 
    # utilizando un muestreo aleatorio o secuencial según el tipo de muestreo indicado.
    def generate_partitions(
        self, 
        sampsize: int, 
        sampling_type: SamplingType
    ) -> List[List[Customer]]:
        partitions: List[List[Customer]] = []
        partition: List[Customer] = []
        
        total_customers = len(Problem.get_problem().get_customers())
        total_partitions = total_customers // sampsize
        
        if total_customers % sampsize != 0:
            total_partitions += 1
            
        print(f"TOTAL DE PARTICIONES: {total_partitions}")
        print("---------------------------------------------------------------")
        
        if sampling_type == SamplingType.RANDOM_SAMPLING:
            j = 0
            customers = list(Problem.get_problem().get_customers())
            
            for i in range(total_partitions):
                while j < (sampsize * (i + 1)) and j < total_customers:
                    pos_element = random.randint(0, len(customers) - 1)
                    partition.append(customers.pop(pos_element))
                    j += 1
                
                print(f"PARTICIÓN {i + 1}: ")
                print(f"TOTAL DE ELEMENTOS DE LA PARTICIÓN {i + 1}: {len(partition)}")
                for customer in partition:
                    print(f"ELEMENTOS DE LA PARTICIÓN {i + 1}: {customer.get_id_customer()}")
                print("---------------------------------------------------------------")
                
                partitions.append(partition)
                partition = []
        
        elif sampling_type == SamplingType.SEQUENTIAL_SAMPLING:
            j = 0
            
            for i in range(total_partitions):
                while j < (sampsize * (i + 1)) and j < total_customers:
                    partition.append(Problem.get_problem().get_customers()[j])
                    j += 1
                
                print(f"PARTICIÓN {i + 1}: ")
                print(f"TOTAL DE ELEMENTOS DE LA PARTICIÓN {i + 1}: {len(partition)}")
                for customer in partition:
                    print(f"ELEMENTOS DE LA PARTICIÓN {i + 1}: {customer.get_id_customer()}")
                print("---------------------------------------------------------------")
                
                partitions.append(partition)
                partition = []
        
        return partitions
    
    # Método que genera elementos iniciales (centroides o medoides) para los clústeres.
    def generate_elements(self, customers: List[Customer]) -> List[int]:
        id_elements: List[int] = []
        total_depots: int = Problem.get_problem().get_total_depots()
        total_customers: int = len(customers)
        counter = total_depots
        
        cost_matrix: np.ndarray = self.initialize_cost_matrix(
            customers, Problem.get_problem().get_depots(), self.distance_type)
        
        print(f"LISTADO DE ELEMENTOS SELECCIONADOS: {id_elements}")
        
        while counter > 0:
            min_value_index = np.argmin(cost_matrix)
            row = min_value_index // cost_matrix.shape[1]
            col = min_value_index % cost_matrix.shape[1]
            
            print(f"FILA SELECCIONADA: {row}")
            print(f"COLUMNA SELECCIONADA: {col}")
            print(f"VALOR SELECCIONADO: {cost_matrix[row, col]}")
            
            id_element = customers[col].get_id_customer()
            id_elements.append(id_element)
        
            print(f"ELEMENTO: {id_element}")
            print(f"LISTADO DE ELEMENTOS ACTUALIZADOS: {id_elements}")
            
            cost_matrix[row, col] = float('inf')
            counter -= 1
        
        print("--------------------------------------------------")
        print("MEDOIDES INICIALES")
        print(id_elements)
        print("--------------------------------------------------")
        
        return id_elements
    
    # Método que obtiene y retorna una lista de los identificadores de los medoides actuales.
    def get_id_medoids(self, medoids: List[Depot]) -> List[int]:
        id_medoids: List[int] = []
        
        for medoid in medoids:
            id_medoids.append(medoid.get_id_depot())
        
        print("--------------------------------------------------")
        print("ID MEDOIDES ACTUALES")
        print("--------------------------------------------------")
        print(id_medoids)
        
        return id_medoids

    # Método que actualiza los clusters asignándoles nuevos elementos y ajustando su demanda 
    # en base a los IDs proporcionados.
    def update_clusters(self, clusters: List[Cluster], id_elements: List[int]):
        for i in range(len(clusters)):
            clusters[i].get_items_of_cluster().append(id_elements[i])
            clusters[i].set_request_cluster(Problem.get_problem().get_request_by_id_customer(id_elements[i]))
            
        print("--------------------------------------------------")
        print("LISTA DE CLUSTERS")
        for cluster in clusters:
            print(f"ID CLUSTER: {cluster.get_id_cluster()}")
            print(f"DEMANDA DEL CLUSTER: {cluster.get_request_cluster()}")
            print(f"ELEMENTOS DEL CLUSTER: {cluster.get_items_of_cluster()}")
        print("--------------------------------------------------")
    
    # Método que verifica si ha habido cambios en las ubicaciones de los medoides 
    # comparando sus coordenadas actuales con las anteriores.
    def verify_medoids(self, old_medoids: List[Depot], current_medoids: List[Depot]) -> bool:
        change: bool = False
        i = 0
        
        while not change and i < len(current_medoids):
            if (old_medoids[i].get_location_depot().get_axis_x() != current_medoids[i].get_location_depot().get_axis_x()) or \
            (old_medoids[i].get_location_depot().get_axis_y() != current_medoids[i].get_location_depot().get_axis_y()):
                change = True
            else:
                i += 1

        print(f"change: {change}")

        return change
       
    # Método que replica los depósitos actuales para crear una nueva lista de depósitos.
    def replicate_depots(self, depots: List[Depot]) -> List[Depot]:
        new_depots: List[Depot] = []
        
        print("--------------------------------------------------")
        print("MEDOIDES/CENTROIDES ACTUALES")

        for i, depot in enumerate(depots):
            depot = Depot()
            depot.set_id_depot(depots[i].get_id_depot())

            axis_x = depots[i].get_location_depot().get_axis_x()
            axis_y = depots[i].get_location_depot().get_axis_y()

            location = Location()
            location.set_axis_x(axis_x)
            location.set_axis_y(axis_y)
            depot.set_location_depot(location)

            fleet = Problem.get_problem().get_depots()[i].get_fleet_depot()
            depot.set_fleet_depot(fleet)

            new_depots.append(depot)

            print("--------------------------------------------------")
            print(f"ID MEDOIDE/CENTROIDE: {new_depots[i].get_id_depot()}")
            print(f"X: {new_depots[i].get_location_depot().get_axis_x()}")
            print(f"Y: {new_depots[i].get_location_depot().get_axis_y()}")
            print(f"CAPACIDAD DE VEHÍCULO: {new_depots[i].get_fleet_depot()[0].get_capacity_vehicle()}")
            print(f"CANTIDAD DE VEHÍCULOS: {new_depots[i].get_fleet_depot()[0].get_count_vehicles()}")

        return new_depots